### PR TITLE
[2.x]Changes in DataSourceService and DataSourceMetadataStorage interface

### PR DESCRIPTION
**Pull request by vamsi-amazon**
### Definitions
* **Datasource** : Datasource class holds basic datasource info(name, connector), storage engine and other objects created specific to a datasource.
* **DatasourceMetadata**:  DatasourceMetadata class carries all the metadata required for a datasource.
* **DatasourceService**: This class acts as interface for datasources and datasource metadata management.
* **DatasourceMetadataStorageService**: This class acts as interface for datasource metadata storage.

### usecases
* CRUD APIs for datasources.
* Analyzer class
    * Fully Qualified Name Resolution requires names of all datasources. [Queries GetAllDatasourceMetadata]
    * DataSourceTable class requires metadata(name, connector) of all datasources for show datasources command. [Queries GetAllDatasourceMetadata]
    * Requires storage engine to implement and execute queries. [Queries GetAllDatasourceMetadata]

#### Datasource bootstrapping during plugin startup [Internal usecase for storage engine creation]
Screen Shot 2023-03-06 at 5 56 34 PM

#### Create datasource [REST API]

Screen Shot 2023-03-06 at 6 09 58 PM

#### Get datasource along with storage engine and other artifacts [Internal use case only]

Screen Shot 2023-03-06 at 6 13 39 PM

#### Get datasource [REST API]

Screen Shot 2023-03-06 at 6 21 26 PM

#### Delete datasource [REST API]

Screen Shot 2023-03-06 at 6 21 44 PM

 
 
  
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/datasource/DataSourceService.java [hunk 1]
```diff
@@ -19,7 +19,7 @@ public interface DataSourceService {
    *
    * @return set of {@link DataSource}.
    */
-  Set<DataSource> getDataSources();
+  Set<DataSourceMetadata> getMaskedDataSourceMetadataSet();
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/datasource/DataSourceService.java:22]: what is masked? could you change the java doc also?

Comment (vamsi-amazon) [reply core/src/main/java/org/opensearch/sql/datasource/DataSourceService.java:22]: Credential info will be removed from the metadata, so masked. Will change the java document.

Comment (penghuo) [reply core/src/main/java/org/opensearch/sql/datasource/DataSourceService.java:22]: why change to return DataSourceMetadata, not sure the reason, but I think DataSourceService has getDataSources method is more natural?

Comment (vamsi-amazon) [reply core/src/main/java/org/opensearch/sql/datasource/DataSourceService.java:22]: Let me cover all the use cases in description and will see if it is required or will revert. 
Show datasources command works with DataSourceService which actually requires datasource metadata but not datasource. 
But with the current information that we provide we can work with getDatasources as well.

**Review by penghuo — COMMENTED** [review]
thanks for the change!
Could you merge to main first? Then backport to 2.x.

**Review by dai-chen — APPROVED** [review]
Thanks for the refactoring! Could you add the detailed design in PR description to a dev doc under https://github.com/opensearch-project/sql/blob/main/docs/dev/index.md#data-sources ?

**Comment by vamsi-amazon** [comment]
https://github.com/opensearch-project/sql/pull/1414
Closing in favour of this backport PR.

**Pull request closed without merging**